File: backend/src/core/Game.ts
import type {
  GameEvent,
  GamePhase,
  Orientation,
  PlacementProgress,
  PlayerSide,
  ShipKind,
  ShipActionRequest,
} from "@shared/index.js";
import { BOARD_SIZE, FLEET_ORDER } from "../config.js";
import { buildFleet } from "../models/ShipFactory.js";
import type { Ship } from "../models/Ship.js";
import { ApiError } from "../api/errors.js";
import { Board } from "./Board.js";
import { Rng } from "./rng.js";
import { getAction } from "./actions/registry.js";

/**
 * Aggregate root: owns the two boards, the rng, the phase, and the
 * placement progress. Mutates itself in response to action requests.
 */
export class Game {
  readonly id: string;
  readonly rng: Rng;
  readonly humanBoard: Board;
  readonly aiBoard: Board;
  readonly placement: PlacementProgress;

  phase: GamePhase = "placement";
  activePlayer: PlayerSide = "human";
  winner?: PlayerSide;

  constructor(id: string, rng: Rng) {
    this.id = id;
    this.rng = rng;
    this.humanBoard = new Board(BOARD_SIZE, "human", buildFleet("human"));
    this.aiBoard = new Board(BOARD_SIZE, "ai", buildFleet("ai"));
    this.placement = { order: [...FLEET_ORDER], nextIndex: 0 };
  }

  // ----- Lookup helpers -------------------------------------------------

  boardFor(side: PlayerSide): Board {
    return side === "human" ? this.humanBoard : this.aiBoard;
  }

  enemyBoardFor(side: PlayerSide): Board {
    return side === "human" ? this.aiBoard : this.humanBoard;
  }

  shipFor(side: PlayerSide, shipId: string): Ship | undefined {
    return this.boardFor(side).ships.find((s) => s.id === shipId);
  }

  // ----- Placement ------------------------------------------------------

  placeHumanShip(kind: ShipKind, anchor: number, orientation: Orientation): void {
    if (this.phase !== "placement") {
      throw new ApiError("WRONG_PHASE", "Ships can only be placed in the placement phase.");
    }
    const expected = this.placement.order[this.placement.nextIndex];
    if (kind !== expected) {
      throw new ApiError(
        "INVALID_PLACE",
        `Expected to place ${expected} next, got ${kind}.`,
      );
    }
    const ship = this.humanBoard.ships.find((s) => s.kind === kind);
    if (!ship) throw new ApiError("INTERNAL", `No ship of kind ${kind} on human board.`);
    try {
      this.humanBoard.placeShip(ship, anchor, orientation);
    } catch (e) {
      throw new ApiError("INVALID_PLACE", (e as Error).message);
    }
    this.placement.nextIndex += 1;
    if (this.placement.nextIndex >= this.placement.order.length) {
      this.phase = "playing";
      this.activePlayer = "human";
      this.resetQuotasFor("human");
    }
  }

  placeHumanRandom(): void {
    if (this.phase !== "placement") {
      throw new ApiError("WRONG_PHASE", "Random placement only allowed in placement phase.");
    }
    // Clear what's there and re-place everything.
    this.humanBoard.clearAllShips();
    this.placement.nextIndex = 0;
    autoPlaceFleet(this.humanBoard, this.rng);
    this.placement.nextIndex = this.placement.order.length;
    this.phase = "playing";
    this.activePlayer = "human";
    this.resetQuotasFor("human");
  }

  resetHumanPlacement(): void {
    if (this.phase !== "placement") {
      throw new ApiError("WRONG_PHASE", "Cannot reset placement once playing.");
    }
    this.humanBoard.clearAllShips();
    this.placement.nextIndex = 0;
  }

  // ----- Action application -------------------------------------------

  applyHumanAction(req: ShipActionRequest): GameEvent[] {
    if (this.phase === "gameover") {
      throw new ApiError("GAME_OVER", "Game has ended.");
    }
    if (this.phase !== "playing") {
      throw new ApiError("WRONG_PHASE", "Actions are only allowed during play.");
    }
    if (this.activePlayer !== "human") {
      throw new ApiError("NOT_YOUR_TURN", "It is not your turn.");
    }
    return this.applyAction("human", req);
  }

  /**
   * Internal action application. Used by both `applyHumanAction` and by the
   * `AIOpponent` which calls this directly under `activePlayer === "ai"`.
   */
  applyAction(side: PlayerSide, req: ShipActionRequest): GameEvent[] {
    const ship = this.shipFor(side, req.shipId);
    if (!ship) throw new ApiError("INVALID_ACTION", `Unknown ship ${req.shipId}.`);
    if (ship.side !== side) {
      throw new ApiError("INVALID_ACTION", "That ship is not yours.");
    }
    if (ship.sunk) {
      throw new ApiError("INVALID_ACTION", `${ship.name} is sunk and cannot act.`);
    }
    if (!ship.supportedActions().includes(req.kind)) {
      throw new ApiError(
        "INVALID_ACTION",
        `${ship.name} does not support action ${req.kind}.`,
      );
    }
    if (!ship.hasQuota(req.kind)) {
      throw new ApiError("NO_QUOTA", `${ship.name} cannot perform ${req.kind} this turn.`);
    }

    const action = getAction(req.kind);
    const ctx = {
      game: this,
      ship,
      ownBoard: this.boardFor(side),
      enemyBoard: this.enemyBoardFor(side),
      request: req,
    };
    action.validate(ctx);
    const events = action.execute(ctx);

    // Check for game-over after every action.
    if (this.enemyBoardFor(side).allShipsSunk()) {
      this.phase = "gameover";
      this.winner = side;
      events.push({ type: "game_over", winner: side });
    }
    return events;
  }

  /**
   * End the human's turn and transition to the AI. Does NOT execute any AI
   * actions; the caller must invoke `ai-step` (one action per call) until the
   * AI has no more quotas, at which point control transitions back to the
   * human automatically via `finishAiTurnIfDone`.
   */
  endHumanTurnTransition(): GameEvent[] {
    if (this.phase !== "playing") {
      throw new ApiError("WRONG_PHASE", "Cannot end turn outside of play.");
    }
    if (this.activePlayer !== "human") {
      throw new ApiError("NOT_YOUR_TURN", "Only the active human turn may be ended this way.");
    }
    // Reset quotas for next time the human plays.
    this.resetQuotasFor("human");
    this.activePlayer = "ai";
    this.resetQuotasFor("ai");
    return [{ type: "turn_started", player: "ai" }];
  }

  /**
   * Transition control back to the human when the AI has nothing left to do.
   * Idempotent; returns the `turn_started: human` event when it actually
   * makes the transition, otherwise returns [].
   */
  finishAiTurnIfDone(): GameEvent[] {
    if (this.phase !== "playing") return [];
    if (this.activePlayer !== "ai") return [];
    if (this.aiHasActionsLeft()) return [];
    this.activePlayer = "human";
    this.resetQuotasFor("human");
    return [{ type: "turn_started", player: "human" }];
  }

  resetQuotasFor(side: PlayerSide): void {
    for (const s of this.boardFor(side).ships) {
      s.resetQuotas();
    }
  }

  // ----- Status helpers -----------------------------------------------

  isOver(): boolean {
    return this.phase === "gameover";
  }

  /** True when the human has at least one usable action across all ships. */
  humanHasActionsLeft(): boolean {
    return this.humanBoard.ships.some(
      (s) => !s.sunk && s.totalActionsLeft() > 0,
    );
  }

  /** True when the AI has at least one usable action across all ships. */
  aiHasActionsLeft(): boolean {
    return this.aiBoard.ships.some(
      (s) => !s.sunk && s.totalActionsLeft() > 0,
    );
  }
}

// ----- Random placement (used by Game and AI) ----------------------------

import { rowCol, shipCells } from "./coords.js";

/**
 * Random rejection-sampling placement. Lives in this file (rather than ai/)
 * because Game uses it for the human "place random" affordance too.
 */
export function autoPlaceFleet(board: Board, rng: Rng): void {
  board.clearAllShips();
  for (const ship of board.ships) {
    let placed = false;
    for (let attempt = 0; attempt < 500 && !placed; attempt++) {
      const orientation: Orientation = rng.nextFloat() < 0.5 ? "horizontal" : "vertical";
      const r = rng.nextInt(board.size);
      const c = rng.nextInt(board.size);
      const cells = shipCells(r, c, ship.length, orientation, board.size);
      if (!cells) continue;
      if (!board.canPlaceCells(cells, ship)) continue;
      // Place via the board to keep cell knowledge in sync.
      const { row, col } = rowCol(cells[0]!, board.size);
      try {
        board.placeShip(ship, cells[0]!, orientation);
        // Anchor in shipCells uses (r,c) directly, which equals positions[0]
        // by construction; we placed via the same coords. Use _row/_col to
        // silence lint about unused.
        void row;
        void col;
        placed = true;
      } catch {
        /* retry */
      }
    }
    if (!placed) {
      // Pathological: start over. Recurse on a clean board.
      board.clearAllShips();
      autoPlaceFleet(board, rng);
      return;
    }
  }
}

File: backend-rs/src/core/game.rs
use crate::api::errors::{ApiError, ApiErrorCode, ApiResult};
use crate::config::{BOARD_SIZE, FLEET_ORDER};
use crate::models::{build_fleet, Ship};
use crate::shared::{
    GameEvent, GamePhase, Orientation, PlacementProgress, PlayerSide, ShipActionRequest, ShipKind,
};

use super::actions::{execute, validate, ActionCtx};
use super::board::Board;
use super::coords::{row_col, ship_cells};
use super::rng::Rng;

/// Aggregate root: owns the two boards, the rng, the phase, and the
/// placement progress.
pub struct Game {
    pub id: String,
    pub rng: Rng,
    pub human_board: Board,
    pub ai_board: Board,
    pub placement: PlacementProgress,
    pub phase: GamePhase,
    pub active_player: PlayerSide,
    pub winner: Option<PlayerSide>,
}

impl Game {
    pub fn new(id: String, rng: Rng) -> Self {
        Self {
            id,
            rng,
            human_board: Board::new(BOARD_SIZE, PlayerSide::Human, build_fleet(PlayerSide::Human)),
            ai_board: Board::new(BOARD_SIZE, PlayerSide::Ai, build_fleet(PlayerSide::Ai)),
            placement: PlacementProgress {
                order: FLEET_ORDER.to_vec(),
                next_index: 0,
            },
            phase: GamePhase::Placement,
            active_player: PlayerSide::Human,
            winner: None,
        }
    }

    pub fn board_for(&self, side: PlayerSide) -> &Board {
        match side {
            PlayerSide::Human => &self.human_board,
            PlayerSide::Ai => &self.ai_board,
        }
    }

    pub fn enemy_board_for(&self, side: PlayerSide) -> &Board {
        match side {
            PlayerSide::Human => &self.ai_board,
            PlayerSide::Ai => &self.human_board,
        }
    }

    pub fn board_for_mut(&mut self, side: PlayerSide) -> &mut Board {
        match side {
            PlayerSide::Human => &mut self.human_board,
            PlayerSide::Ai => &mut self.ai_board,
        }
    }

    pub fn ship_idx_for(&self, side: PlayerSide, ship_id: &str) -> Option<usize> {
        self.board_for(side).find_ship_idx_by_id(ship_id)
    }

    // ----- Placement ------------------------------------------------------

    pub fn place_human_ship(
        &mut self,
        kind: ShipKind,
        anchor: u32,
        orientation: Orientation,
    ) -> ApiResult<()> {
        if self.phase != GamePhase::Placement {
            return Err(ApiError::new(
                ApiErrorCode::WrongPhase,
                "Ships can only be placed in the placement phase.",
            ));
        }
        let expected = self.placement.order[self.placement.next_index as usize];
        if kind != expected {
            return Err(ApiError::new(
                ApiErrorCode::InvalidPlace,
                format!("Expected to place {:?} next, got {:?}.", expected, kind),
            ));
        }
        let Some(idx) = self.human_board.find_ship_idx_by_kind(kind) else {
            return Err(ApiError::new(
                ApiErrorCode::Internal,
                format!("No ship of kind {:?} on human board.", kind),
            ));
        };
        self.human_board
            .place_ship(idx, anchor, orientation)
            .map_err(|e| ApiError::new(ApiErrorCode::InvalidPlace, e))?;
        self.placement.next_index += 1;
        if self.placement.next_index as usize >= self.placement.order.len() {
            self.phase = GamePhase::Playing;
            self.active_player = PlayerSide::Human;
            self.reset_quotas_for(PlayerSide::Human);
        }
        Ok(())
    }

    pub fn place_human_random(&mut self) -> ApiResult<()> {
        if self.phase != GamePhase::Placement {
            return Err(ApiError::new(
                ApiErrorCode::WrongPhase,
                "Random placement only allowed in placement phase.",
            ));
        }
        self.human_board.clear_all_ships();
        self.placement.next_index = 0;
        auto_place_fleet(&mut self.human_board, &mut self.rng);
        self.placement.next_index = self.placement.order.len() as u32;
        self.phase = GamePhase::Playing;
        self.active_player = PlayerSide::Human;
        self.reset_quotas_for(PlayerSide::Human);
        Ok(())
    }

    pub fn reset_human_placement(&mut self) -> ApiResult<()> {
        if self.phase != GamePhase::Placement {
            return Err(ApiError::new(
                ApiErrorCode::WrongPhase,
                "Cannot reset placement once playing.",
            ));
        }
        self.human_board.clear_all_ships();
        self.placement.next_index = 0;
        Ok(())
    }

    // ----- Action application -------------------------------------------

    pub fn apply_human_action(&mut self, req: &ShipActionRequest) -> ApiResult<Vec<GameEvent>> {
        if self.phase == GamePhase::Gameover {
            return Err(ApiError::new(ApiErrorCode::GameOver, "Game has ended."));
        }
        if self.phase != GamePhase::Playing {
            return Err(ApiError::new(
                ApiErrorCode::WrongPhase,
                "Actions are only allowed during play.",
            ));
        }
        if self.active_player != PlayerSide::Human {
            return Err(ApiError::new(
                ApiErrorCode::NotYourTurn,
                "It is not your turn.",
            ));
        }
        self.apply_action(PlayerSide::Human, req)
    }

    /// Internal action application. Used by both human actions and the AI.
    pub fn apply_action(
        &mut self,
        side: PlayerSide,
        req: &ShipActionRequest,
    ) -> ApiResult<Vec<GameEvent>> {
        let ship_idx = self.ship_idx_for(side, &req.ship_id).ok_or_else(|| {
            ApiError::new(
                ApiErrorCode::InvalidAction,
                format!("Unknown ship {}.", req.ship_id),
            )
        })?;
        // Pre-checks on a shared ref before splitting borrows.
        {
            let ship: &Ship = &self.board_for(side).ships[ship_idx];
            if ship.side != side {
                return Err(ApiError::new(
                    ApiErrorCode::InvalidAction,
                    "That ship is not yours.",
                ));
            }
            if ship.sunk() {
                return Err(ApiError::new(
                    ApiErrorCode::InvalidAction,
                    format!("{} is sunk and cannot act.", ship.name()),
                ));
            }
            if !ship.supported_actions().contains(&req.kind) {
                return Err(ApiError::new(
                    ApiErrorCode::InvalidAction,
                    format!("{} does not support action {:?}.", ship.name(), req.kind),
                ));
            }
            if !ship.has_quota(req.kind) {
                return Err(ApiError::new(
                    ApiErrorCode::NoQuota,
                    format!(
                        "{} cannot perform {:?} this turn.",
                        ship.name(),
                        req.kind
                    ),
                ));
            }
        }

        // Split mutable borrows of the two boards.
        let (own_board, enemy_board) = match side {
            PlayerSide::Human => {
                let (a, b) = split_boards(&mut self.human_board, &mut self.ai_board);
                (a, b)
            }
            PlayerSide::Ai => {
                let (b, a) = split_boards(&mut self.human_board, &mut self.ai_board);
                (a, b)
            }
        };

        let mut ctx = ActionCtx {
            side,
            ship_idx,
            own_board,
            enemy_board,
            request: req,
        };
        validate(req.kind, &ctx)?;
        let mut events = execute(req.kind, &mut ctx)?;

        // Check for game-over.
        let enemy_all_sunk = self.enemy_board_for(side).all_ships_sunk();
        if enemy_all_sunk {
            self.phase = GamePhase::Gameover;
            self.winner = Some(side);
            events.push(GameEvent::GameOver { winner: side });
        }
        Ok(events)
    }

    pub fn end_human_turn_transition(&mut self) -> ApiResult<Vec<GameEvent>> {
        if self.phase != GamePhase::Playing {
            return Err(ApiError::new(
                ApiErrorCode::WrongPhase,
                "Cannot end turn outside of play.",
            ));
        }
        if self.active_player != PlayerSide::Human {
            return Err(ApiError::new(
                ApiErrorCode::NotYourTurn,
                "Only the active human turn may be ended this way.",
            ));
        }
        self.reset_quotas_for(PlayerSide::Human);
        self.active_player = PlayerSide::Ai;
        self.reset_quotas_for(PlayerSide::Ai);
        Ok(vec![GameEvent::TurnStarted {
            player: PlayerSide::Ai,
        }])
    }

    pub fn finish_ai_turn_if_done(&mut self) -> Vec<GameEvent> {
        if self.phase != GamePhase::Playing {
            return Vec::new();
        }
        if self.active_player != PlayerSide::Ai {
            return Vec::new();
        }
        if self.ai_has_actions_left() {
            return Vec::new();
        }
        self.active_player = PlayerSide::Human;
        self.reset_quotas_for(PlayerSide::Human);
        vec![GameEvent::TurnStarted {
            player: PlayerSide::Human,
        }]
    }

    pub fn reset_quotas_for(&mut self, side: PlayerSide) {
        for s in &mut self.board_for_mut(side).ships {
            s.reset_quotas();
        }
    }

    pub fn is_over(&self) -> bool {
        self.phase == GamePhase::Gameover
    }

    pub fn human_has_actions_left(&self) -> bool {
        self.human_board
            .ships
            .iter()
            .any(|s| !s.sunk() && s.total_actions_left() > 0)
    }

    pub fn ai_has_actions_left(&self) -> bool {
        self.ai_board
            .ships
            .iter()
            .any(|s| !s.sunk() && s.total_actions_left() > 0)
    }
}

fn split_boards<'a>(a: &'a mut Board, b: &'a mut Board) -> (&'a mut Board, &'a mut Board) {
    (a, b)
}

/// Random rejection-sampling placement. Iterative outer retry loop to avoid
/// recursion (the TS version recurses on failure).
pub fn auto_place_fleet(board: &mut Board, rng: &mut Rng) {
    'retry: loop {
        board.clear_all_ships();
        let n = board.ships.len();
        for i in 0..n {
            let length = board.ships[i].length();
            let size = board.size;
            let mut placed = false;
            for _ in 0..500 {
                let orientation = if rng.next_float() < 0.5 {
                    Orientation::Horizontal
                } else {
                    Orientation::Vertical
                };
                let r = rng.next_int(size) as i32;
                let c = rng.next_int(size) as i32;
                let Some(cells) = ship_cells(r, c, length, orientation, size) else {
                    continue;
                };
                if !board.can_place_cells(&cells, Some(i)) {
                    continue;
                }
                let (anchor_row, anchor_col) = row_col(cells[0], size);
                let _ = (anchor_row, anchor_col);
                if board.place_ship(i, cells[0], orientation).is_ok() {
                    placed = true;
                    break;
                }
            }
            if !placed {
                continue 'retry;
            }
        }
        break;
    }
}
